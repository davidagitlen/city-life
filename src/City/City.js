import React, { Component } from 'react';
import './City.scss';
import { fetchUrbanAreas, fetchCityScores, findAdditionalData, fetchAdditionalData } from '../util/apiCalls';

class City extends Component{
  constructor(){
    super();
    this.state = {
      city : '',
      cities : [],
      error : ''
    };
  }

  componentDidMount = async () => {
    try {
      const urbanAreas = await fetchUrbanAreas();
      const urbanAreasNames = urbanAreas._links['ua:item'];
      this.setState({cities: urbanAreasNames}, () => {console.log(urbanAreasNames)});
    } catch ({ message }) {
      this.setState({error : message}, () => {console.error(message)});
    }
  }

  handleCitySelection = e => {
    this.setState({[e.target.name]:e.target.value})
  }

  handleSubmitCity = async (e) => {
    e.preventDefault();
    this.getCityScores();
    this.handleAdditionalData();
  }

  getCityScores = async () => {
    console.log(this.state.city.toLowerCase())
    const citySnippet = this.state.city.toLowerCase().replace(/,|\./g, '').replace(/\s/g, '-');
    try {
      const city = await fetchCityScores(`https://api.teleport.org/api/urban_areas/slug:${citySnippet}/`);
      const cityScores = city.categories;
      console.log(cityScores);
    } catch ({ message }) {
      this.setState({ error: message }, () => { console.error(message) });
    }
  }

  getAdditionalCityData = async () => {
    const originalName = this.state.city;
    const cityName = this.formatCityName(this.state.city).replace(/,|\./, '')
    try {
      const city = await findAdditionalData(cityName, originalName);
      return city;
    } catch ({ message }) {
      this.setState({ error: message}, () => {console.error(message)});
    }
  }

  formatCityName = (name) => {
    switch(name) {
      case 'San Francisco Bay Area' :
        return 'San Francisco';
      case 'Portland, ME' :
        return 'Portland';
      case 'Portland, OR' :
        return 'Portland';
      case 'Birmingham, AL' :
        return 'Birmingham';
      case 'Minneapolis-Saint Paul' :
        return 'Minneapolis';
      default :
        return name; 
    }
  }

  handleAdditionalData = async () => {
    const href = await this.getAdditionalCityData();
    try {
      const additionalData = await fetchAdditionalData(href);
      console.log('additionalData return', additionalData)
      const formattedData = {
        population: additionalData.population,
        latitude: additionalData.location.latlon.latitude,
        longitude: additionalData.location.latlon.longitude,
        fullName: additionalData.full_name,
        timeZone: additionalData._links['city:timezone'].name,
        country: additionalData._links['city:country'].name
      }
      console.log(formattedData)
      return formattedData;
    } catch ({ message }) {
      this.setState({ error : message});
    }
  }


  
  render() {
    console.log(this.props)
    const cities = this.state.cities;
    const cityNames = cities.length ? cities.map(city => <option key={city.name} value={city.name} />) : null;

    return(
      <div>
        <form>
          <input 
            list='city-selection'
            name='city'
            type='text' 
            value={this.state.city}
            placeholder='Select a City'
            autoComplete='off'
            onChange={this.handleCitySelection}
            />
          <datalist id='city-selection'>
            {cityNames}
          </datalist>
          <button type='submit' onClick={this.handleSubmitCity}></button>
        </form>
      </div>
    )
  }
}

export default City;