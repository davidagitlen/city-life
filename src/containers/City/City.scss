.City {
  position: relative;
  background: ghostwhite;
  height: 280px;
  width: 100%;

  & div{
    position: relative;
    overflow: hidden;
  }

  & img {
    height: 200px;
    margin: 15px 0px 15px 0px;
    position: relative;
    max-width: 100%;

    &#circle {
      -webkit-clip-path: circle(75px at 50% 50%);
      clip-path: circle(100px at 50% 50%);
    }

  }

  & p {
    overflow: wrap;
  }
}