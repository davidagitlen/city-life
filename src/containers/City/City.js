import React from 'react';
import './City.scss';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

export const City = (props) => {
  const { cityInfoReducer: cityData, ordinal } = props;
  const { fullName } = cityData.details[ordinal];
  const { web } = cityData.images[ordinal].images;
  const linkName = fullName ? fullName.split(',')[0] : null;
  return(
    <div className='City'>
      <Link to={`details/${linkName}`} className='link'>
        <div>
          <img id='circle' src={web} alt=''/>
        </div>
        <p>{fullName}</p>
      </Link>
    </div>
  )
}

export const mapStateToProps = state => ({
  ...state
})

export default connect(mapStateToProps)(City);

City.propTypes = {
  cityInfo: PropTypes.object,
  dispatch: PropTypes.func,
  ordinal: PropTypes.string
}