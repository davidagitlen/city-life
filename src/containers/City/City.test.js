import React from 'react';
import { City, mapStateToProps }from './City';
import { shallow } from 'enzyme';

describe('City', () => {

  it('should match the snapshot with data passed in correctly', () => {
    const mockData = {
      ordinal: 'one',
      cityInfo : {
        one: {
          details: {
            fullName: 'Dorvord, Gordletan'
          },
          images: {
            images: {
              web: 'http://blah.com'
            }
          }
        },
        two: {
          details: {
            fullName: 'Brindanda, DelBla;bey'
          },
          images: {
            images: {
              web: 'http://blah.com'
            }
          }
        }
      }
    };
    const wrapper = shallow(
      <City 
        {...mockData}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {

    it('should return an object with appropriate data', () => {

      const mockCityInfo = {
        one: {
          details: {
            population: 1,
            latitude: 10,
            longitude: 10,
            fullName: 'Denver, Colorado!',
            timeZone: 'Denver!'
          },
          images: {
            attribution: {},
            images: {
              mobile: 'http://blurg.com',
              web: 'http://blurg.com/web'
            }
          },
          scores: [{ color: 'blue', name: 'Score', score_out_of_10: 10 }]
        },
        two: {
          details: {
            population: 2,
            latitude: 20,
            longitude: 20,
            fullName: 'Boulder, Colorado!',
            timeZone: 'Denver!'
          },
          images: {
            attribution: {},
            images: {
              mobile: 'http://blurg.com',
              web: 'http://blurg.com/web'
            }
          },
          scores: [{ color: 'red', name: 'Score', score_out_of_10: 10 }]
        }
      };

      const mockState = {
        cityInfo: mockCityInfo,
        unrealData: 'blergh!'
      };

      const expected = {
        cityInfo: mockCityInfo
      };

      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
  });
 
});