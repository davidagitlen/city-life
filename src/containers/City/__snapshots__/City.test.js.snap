// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`City should match the snapshot with data passed in correctly 1`] = `
<div
  className="City"
>
  <Link
    className="link"
    to="details/Dorvord"
  >
    <div>
      <img
        alt=""
        id="circle"
        src="http://blah.com"
      />
    </div>
    <p>
      Dorvord, Gordletan
    </p>
  </Link>
</div>
`;
